### PR TITLE
code for merging EPIC KITCHENS and EPIC SOUNDS datasets, organized by timestamp and customizable for data added

### PR DESCRIPTION
**Pull request by laurenx1**
functions that take in the EPIC KITCHENS and EPIC SOUNDS CSVs, creates a new CSV that organizes the data in these 2 CSVs by timestamp. The data put into this new csv is completely customizable, for example I could take any combination of the narration_id, narration, verb, verb_class, noun, noun_class, all_nouns, all_noun_classes columns from EPIC KITCHENS, and any combination of description, narration_id, class, class_id from EPIC SOUNDS. Also, the orientation of the timestamp can be changed, so that the intervals of start and stop fit whatever needs the user has. 

Ultimately, this code allows for better organization and visualization for the correlation of audio and visual data.

### CODE REVIEW

**Comment by iranroman** [comment]
Thank you Lauren, I'll review your code and make suggestions.

@beasteers feel free to also carry out a code review

After that we will merge with main.

**Comment by iranroman** [comment]
Thanks Lauren for bringing this Repo to life. 

I updated the README to reflect some of the main functionalities that we want. Please pull from main and let me know what you think.

**Comment by beasteers** [comment]
I was thinking about how we want to have the API designed so I sketched out some alterations: (sorry I got a little carried away)

First, I think we should have explicit functions for loading each csv. For convenience, they take the root of each annotation repo (but can handle an explicit filename in case someone changed the filename/structure).

```python

def load_epic_kitchens_actions_csv(epic_kitchens_root_dir, splits=('train', 'validation')):
    '''Load the EPIC Kitchens action annotations (for multiple splits)'''
    # handle different filename inputs
    if os.path.isfile(epic_kitchens_root_dir):  # someone passed in a specific file
        filenames = [epic_kitchens_root_dir]
    else:  # assume the default epic kitchens annotation file structure
        # use the splits parameter to select which CSVs to load.
        filenames = [
            os.path.join(epic_kitchens_root_dir, f'EPIC_100_{split}.csv')
            for split in splits
        ]

    # load multiple csv files
    df = pd.concat([pd.read_csv(filename) for filename in filenames])
    # parse the start and stop times into datetime objects
    df['start_timestamp'] = pd.to_datetime(df.start_timestamp, format="%H:%M:%S.%f")
    df['stop_timestamp'] = pd.to_datetime(df.stop_timestamp, format="%H:%M:%S.%f")
    return df
    

def load_epic_sounds_events_csv(epic_sounds_root_dir, splits=('train', 'validation')):
    # handle different filename inputs
    if os.path.isfile(epic_sounds_root_dir):  # someone passed in a specific file
        filenames = [epic_sounds_root_dir]
    else:  # assume the default epic kitchens annotation file structure
        # use the splits parameter to select which CSVs to load.
        filenames = [
            os.path.join(epic_sounds_root_dir, f'EPIC_100_{split}.csv')
            for split in splits
        ]

    # load multiple csv files
    df = pd.concat([pd.read_csv(filename) for filename in filenames])
    # parse the start and stop times into datetime objects
    df['start_timestamp'] = pd.to_datetime(df.start_timestamp)
    df['stop_timestamp'] = pd.to_datetime(df.stop_timestamp)
    return df


def load_epic_kitchens_video_info(epic_kitchens_root_dir):
    # handle different filename inputs
    if os.path.isfile(filename):  # someone passed in a specific file
        filename = epic_kitchens_root_dir
    else:  # assume the default epic kitchens annotation file structure
        filename = os.path.join(epic_kitchens_root_dir, 'EPIC_100_video_info.csv')
    
    # load the video info csv
    df = pd.read_csv(filename)
    # use the video_id as the index so we can easily lookup the video (e.g. df.loc[video_id])
    df = df.set_index('video_id')
    return df
```

I added `load_epic_kitchens_video_info` which will load the explicit video durations (as well as fps) which is more reliable than using timestamps from the action/sound csvs.

----

Then for the top-level function - realistically, if you were running this, you'd probably want to run it across all videos:

```python

def run(
        epic_kitchens_root_dir, 
        epic_sounds_root_dir, 
        output_dir='epic-kitchens-united', 
        action_columns=['narration_id','verb','all_nouns','verb_class','all_noun_classes'],
        sound_columns=['annotation_id','class','class_id'],
        video_ids=None,
        time_delta=1
):
    # load each of the datasets
    video_info_df = load_epic_kitchens_video_info(epic_kitchens_root_dir)
    action_df = load_epic_kitchens_actions_csv(epic_kitchens_root_dir)
    sound_df = load_epic_sounds_events_csv(epic_sounds_root_dir)

    # if no video ids were provided, create for all videos
    if video_ids is None:
        video_ids = action_df.video_id.unique()

    # generate CSVs for each video
    for video_id in tqdm.tqdm(video_ids):

        # create the filename for the video
        new_filename = os.path.join(output_dir, f'{video_id}.csv')

        # get the duration from the video info file
        duration = video_info_df.loc[video_id].duration

        # create the time csv
        df = make_time_csv(new_filename, time_delta, duration)
        
        # merge columns from each annotation csv
        df = add_columns(df, time_delta, action_df[action_df.video_id == video_id], action_columns)
        df = add_columns(df, time_delta, sound_df[sound_df.video_id == video_id], sound_columns)
        
        # save the csv
        df.to_csv(new_filename)

```

File: epic_kitchens_united/core.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        vid_id (str): the video id, ex. "P01_103"
+
+    Returns: 
+        (DataFrame): the original df filtered by vid_id
+    """
+    return df[df['video_id'] == vid_id]
+
+
+
+def make_time_csv(df, dt, new_filepath, max_t):
```
Comment (beasteers) [inline epic_kitchens_united/core.py:60]: currently this function writes the time csv out to file, but maybe that's not necessary? we could maybe simplify this to create the dataframe in memory.

```python
def make_time_csv(time_delta, duration):
    # convert time_delta to seconds
    time_delta = time_delta/1000
    # create the array of times from 0 to the video duration, 
    # with an increment of time_delta seconds
    start_time = np.arange(0, dt_to_seconds(duration), time_delta)
    # convert from seconds (as a float) to a timedelta array 
    # (just learned you can do it like this :P)
    start_time = start_time * timedelta(seconds=1)
    # create the dataframe directly
    return pd.DataFrame({
        'start_timestamp_inclusive': start_time,
        'stop_timestamp_exclusive': start_time + timedelta(seconds=time_delta),
    })
```

File: epic_kitchens_united/core.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+import pandas as pd
+import csv
+from datetime import datetime
+import os
+import math
+from datetime import timedelta
+import tqdm
+
+
+def time_to_seconds(timestamp):
```
Comment (beasteers) [inline epic_kitchens_united/core.py:12]: this function is unused

**Comment by iranroman** [comment]
Thank you @beasteers. I like your ideas to enhance this API a lot. 

```

# assuming the the two annotation repos are in the current directory
python main.py

# customize like this
python main.py epic-kitchens-100-annotations epic-sounds-annotations \
    --action-columns '["narration_id","verb"]' \
    --sound-columns '["annotation_id","class"]' \
    --video-ids '["P01_01","P01_02"]'
```

will the assumed time-step be `1` (second) if not specified? What if it were to be specified?

**Comment by beasteers** [comment]
yep good questions. The 1 second was arbitrary, but maybe 0.1 seconds would be a better default? Idk what you all think. 

Here's the updated usage with all arguments.

```python
python main.py epic-kitchens-100-annotations epic-sounds-annotations output \
    --time-delta 0.1 \
    --action-columns '["narration_id","verb"]' \
    --sound-columns '["annotation_id","class"]' \
    --video-ids '["P01_01","P01_02"]'
```
Wondering if it makes more sense to have the output directory as the first argument? Depends on what a user is more likely to override.

**Pull request merged**